### PR TITLE
properly escape single quotes values in PK and RK Uri paths

### LINKED ISSUE #21635: [BUG] Single quote in PartitionKey not serialized properly in PATCH entity
**Describe the bug**
When the Table entity partition key contains single quotes, the PATCH operation (where the PK is part of the URL path, not query string) fails. The single quote is URL encoded as `%27`. The legacy SDK escapes it with two single quotes.

Found while trying to fix https://github.com/Azure/Azurite/issues/790.

**Expected behavior**
It should escape the single quote in a way that Table storage expects.

**Actual behavior (include Exception or Stack Trace)**
```
Unhandled exception. Azure.RequestFailedException: Bad Request - Error in query syntax.
RequestId:6539ca23-b002-0045-4051-5a7235000000
Time:2021-06-05T21:29:07.6478169Z
Status: 400 (Bad Request)
ErrorCode: InvalidInput

Content:
{"odata.error":{"code":"InvalidInput","message":{"lang":"en-US","value":"Bad Request - Error in query syntax.\nRequestId:6539ca23-b002-0045-4051-5a7235000000\nTime:2021-06-05T21:29:07.6478169Z"}}}

Headers:
Transfer-Encoding: chunked
Server: Windows-Azure-Table/1.0,Microsoft-HTTPAPI/2.0
x-ms-request-id: 6539ca23-b002-0045-4051-5a7235000000
x-ms-client-request-id: 69ef5d19-5496-4a12-a0f5-3baa9a331b28
x-ms-version: REDACTED
X-Content-Type-Options: REDACTED
Date: Sat, 05 Jun 2021 21:29:07 GMT
Content-Type: application/json; odata=minimalmetadata; streaming=true; charset=utf-8

   at Azure.Data.Tables.TableRestClient.MergeEntity(String table, String partitionKey, String rowKey, Nullable`1 timeout, String ifMatch, IDictionary`2 tableEntityProperties, QueryOptions queryOptions, CancellationToken cancellationToken)
   at Azure.Data.Tables.TableClient.UpsertEntity[T](T entity, TableUpdateMode mode, CancellationToken cancellationToken)
   at ConsoleApp1.Program.Main(String[] args) in C:\Users\joelv\Desktop\ConsoleApp1\ConsoleApp1\Program.cs:line 31
   at ConsoleApp1.Program.<Main>(String[] args)
```

**To Reproduce**
```csharp
using System;
using System.Linq;
using System.Threading.Tasks;
using Azure;
using Azure.Data.Tables;
using Microsoft.WindowsAzure.Storage;
using Microsoft.WindowsAzure.Storage.Table;

namespace ConsoleApp1
{
    class Program
    {
        const string ConnectionString = "UseDevelopmentStorage=true";
        const string TableName = "azuritetest";

        static async Task Main(string[] args)
        {
            {
                var account = CloudStorageAccount.Parse(ConnectionString);
                var tableServiceClient = account.CreateCloudTableClient();
                var table = tableServiceClient.GetTableReference(TableName);
                await table.CreateIfNotExistsAsync();
                await table.ExecuteAsync(TableOperation.InsertOrMerge(new Microsoft.WindowsAzure.Storage.Table.TableEntity("Isn't this a nasty PK", "rk")));
            }

            {
                var tableServiceClient = new TableServiceClient(ConnectionString);
                var tableClient = tableServiceClient.GetTableClient(TableName);
                await tableClient.CreateIfNotExistsAsync();
                tableClient.UpsertEntity(new Azure.Data.Tables.TableEntity("Isn't this a nasty PK", "rk"));
            }
        }
    }
}
```

**Environment:**
 - Name and version of the Library package used: Azure.Data.Tables 12.0.0-alpha.20210603.1
 - Hosting platform or OS and .NET runtime version (`dotnet --info` output for .NET Core projects): .NET SDK 5.0.300
 - IDE and version : VS 16.10

### PR DESCRIPTION
**Pull request by christothes**
fixes #21635

### CODE REVIEW

File: sdk/tables/Azure.Data.Tables/src/TableOdataFilter.cs [hunk 1]
```diff
@@ -80,5 +80,9 @@ public static string Create(FormattableString filter)
 
             return string.Format(CultureInfo.InvariantCulture, filter.Format, args);
         }
+
+        internal static string escapeStringValue(string s) => s.Replace("'", "''");
```
Comment (pakrym) [inline sdk/tables/Azure.Data.Tables/src/TableOdataFilter.cs:84]: nit: casing

File: sdk/tables/Azure.Data.Tables/src/TableOdataFilter.cs [hunk 2]
```diff
@@ -80,5 +80,9 @@ public static string Create(FormattableString filter)
...
+        internal static StringBuilder escapeStringValue(StringBuilder s) => s.Replace("'", "''");
+        internal static string escapeStringValue(char s) => s.ToString().Replace("'", "''");
```
Comment (pakrym) [inline sdk/tables/Azure.Data.Tables/src/TableOdataFilter.cs:86]: nit: why do you hate perf? :)

**Review by pakrym — APPROVED** [review]

File: sdk/tables/Azure.Data.Tables/tests/TableClientLiveTests.cs [hunk 3]
```diff
@@ -27,6 +27,31 @@ public class TableClientLiveTests : TableServiceLiveTestsBase
             endpointType /* To record tests, add this argument, RecordedTestMode.Record */)
         { }
 
+        [RecordedTest]
+        public async Task UpsertAndQueryWithSingleQuoteNames([Values(true, false)] bool expressionQuery)
+        {
+            List<TableEntity> entityResults;
+            string partitionKeyValue = "PartitionWith'singleQuote";
```
Comment (pakrym) [inline sdk/tables/Azure.Data.Tables/tests/TableClientLiveTests.cs:34]: What about a double quote?

Comment (christothes) [reply sdk/tables/Azure.Data.Tables/tests/TableClientLiveTests.cs:34]: double quote or 2 single quotes? The spec only mentions single quotes - https://docs.microsoft.com/en-us/rest/api/storageservices/querying-tables-and-entities#single-quote-

Comment (pakrym) [reply sdk/tables/Azure.Data.Tables/tests/TableClientLiveTests.cs:34]: Two single quotes.

Comment (christothes) [reply sdk/tables/Azure.Data.Tables/tests/TableClientLiveTests.cs]: Updated the test with two quotes just to verify.

File: sdk/tables/Azure.Data.Tables/src/TableOdataFilter.cs [hunk 4]
```diff
@@ -80,5 +80,9 @@ public static string Create(FormattableString filter)
 
             return string.Format(CultureInfo.InvariantCulture, filter.Format, args);
         }
+
+        internal static string EscapeStringValue(string s) => s.Replace("'", "''");
+        internal static StringBuilder EscapeStringValue(StringBuilder s) => s.Replace("'", "''");
+        internal static string EscapeStringValue(char s) => s.ToString().Replace("'", "''");
```
Comment (jsquire) [inline sdk/tables/Azure.Data.Tables/src/TableOdataFilter.cs:86]: Should we consider short-circuiting this if `s` is not the single quote character?   Maybe something like:

```csharp
internal static string EscapeStringValue(char s) => 
    s switch
    {
        _ when s == '\'' => "''",
        _ => s.ToString()
    };
```

**Review by jsquire — APPROVED** [review]

**Pull request merged**